### ISSUE TITLE
Circular dependencies handling

### ISSUE THREAD

**Issue by merapi** [GoogleForCreators/web-stories-wp#1290]
## Feature Description

We now know that circular dependencies can break stuff (this time is was Storybook static preview).

Now we want to know: Is there a way to detect it before it happens?

Small test case setup (most/less obvious cases): https://github.com/google/web-stories-wp/commit/0ab18234f60474782052272877ecdb01a11a382b

Options for detecting CD:
1. [`eslint-plugin-import/no-cycle`](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-cycle.md) rule
2. Tools like [madge](https://github.com/pahen/madge) or [dpdm](https://github.com/acrazing/dpdm)
3. [`circular-dependency-plugin`](https://github.com/aackerman/circular-dependency-plugin) webpack plugin

From my tests, there is no way to detect CD that will break our app. We can just detect all CD.
Here is how the output looks like:
Eslint plugin: 
![image](https://user-images.githubusercontent.com/21036224/79865060-d6ea2600-83da-11ea-9c8a-56919582c862.png)
Madge: 
![image](https://user-images.githubusercontent.com/21036224/79865087-e36e7e80-83da-11ea-833d-226df0729129.png)
Dpdm: 
![image](https://user-images.githubusercontent.com/21036224/79865114-ee291380-83da-11ea-8c5a-1e4286a3e6e4.png)

Right now we have ~6X of them, let's look at one:

```
/web-stories-wp/assets/src/edit-story/elements/media/index.js
  25:1  error  Dependency cycle via ../../elements:20=>./image:31  import/no-cycle
```
![cd1](https://user-images.githubusercontent.com/21036224/79863376-f29ffd00-83d7-11ea-88da-372928663b0e.gif)

But is it causing any harm? Right now - no, as shown below, if you try to use the import immediately it will break the app, if you use it in function (later) it will be okay (like it is right now).
![useimport](https://user-images.githubusercontent.com/21036224/79863831-b8832b00-83d8-11ea-9475-9cf9ed6dd349.gif)

Can we fix it? Probably.
Is it easy? Not at all. We have too many index.js files and mutual dependencies.

Of course, with more E2E testing and some basic Storybook testing, we can detect the error at runtime.

Small note about eslint rule:
`no-cycle` is _comparatively computationally expensive_ so we probably want to use it only during git hooks:
`npm run lint:js -- --rule 'import/no-cycle: [2, { maxDepth: 3 }]'`
_maxDepth is optional, default to Infinity_

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

**Comment by miina** [comment]
For information:
The playground is currently broken due to this import specifically added in https://github.com/google/web-stories-wp/commit/ae4a387ac6607fb7c44b6dddf9ab58c77afd90a0:
https://github.com/google/web-stories-wp/blob/7025fab42f299f2b30f3e9b06e4069396e02e997/assets/src/dashboard/utils/index.js#L31

From discussion with @swissspidy -- this import causes webpack importing things in a different order which might have uncovered new issues with dependencies.

**Comment by swissspidy** [comment]
^ the above appears to be fixed by #2614.
